smack-extensions/src/main/java/org/jivesoftware/smackx/avatar/UserAvatarManager.java

```diff
@@ -0,0 +1,318 @@
+        ServiceDiscoveryManager serviceDiscoveryManager = ServiceDiscoveryManager.getInstanceFor(connection());
+        List<Item> items = serviceDiscoveryManager.discoverItems(connection().getUser().asBareJid(), nodeName)
+                .getItems();
+        return items != null && items.size() > 0;
+    }
+
+    /**
+     * Get the data node.
+     *
+     * @return the data node
+     * @throws NoResponseException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     * @throws XMPPErrorException
+     */
+    public LeafNode getDataNode()
+            throws NoResponseException, NotConnectedException, InterruptedException, XMPPErrorException, PubSubException.NotALeafNodeException {
+        return pepManager.getPepPubSubManager().getOrCreateLeafNode(DATA_NAMESPACE);
+    }
+
+    /**
+     * Get the metadata node.
+     *
+     * @return the metadata node
+     * @throws NoResponseException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     * @throws XMPPErrorException
+     */
+    public LeafNode getMetadataNode()
+            throws NoResponseException, NotConnectedException, InterruptedException, XMPPErrorException, PubSubException.NotALeafNodeException {
+        return pepManager.getPepPubSubManager().getOrCreateLeafNode(METADATA_NAMESPACE);
+    }
+
+    /**
+     * Publish avatar data.
+     *
+     * @param data
+     * @param itemId
+     * @throws NoResponseException
+     * @throws NotConnectedException
+     * @throws XMPPErrorException
+     * @throws InterruptedException
+     */
+    public void publishAvatarData(byte[] data, String itemId)
+            throws NoResponseException, NotConnectedException, XMPPErrorException, InterruptedException, PubSubException.NotALeafNodeException {
+        DataExtension dataExtension = new DataExtension(data);
+        getDataNode().publish(new PayloadItem<>(itemId, dataExtension));
+    }
+
+    /**
+     * Publish avatar data.
+     *
+     * @param data
+     * @return created item id
+     * @throws NoResponseException
+     * @throws NotConnectedException
+     * @throws XMPPErrorException
+     * @throws InterruptedException
+     */
+    public String publishAvatarData(byte[] data)
+            throws NoResponseException, NotConnectedException, XMPPErrorException, InterruptedException, PubSubException.NotALeafNodeException {
+        String itemId = Base64.encodeToString(SHA1.bytes(data));
+        publishAvatarData(data, itemId);
+        return itemId;
+    }
+
+    /**
+     * Publish avatar metadata.
+     *
+     * @param itemId
+     * @param info
+     * @param pointers
+     * @throws NoResponseException
+     * @throws XMPPErrorException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     */
+    public void publishAvatarMetadata(String itemId, MetadataInfo info, List<MetadataPointer> pointers)
```

Comment:
I'd say, publishing metadata should be done automatically when an avatar is uploaded.

Reply:
Sounds sensible.